asm/load_store.go

```diff
@@ -26,10 +26,65 @@ const (
 	MemMode Mode = 0x60
 	// MemSXMode - load from memory, sign extension
 	MemSXMode Mode = 0x80
-	// XAddMode - add atomically across processors.
-	XAddMode Mode = 0xc0
+	// AtomicMode - add atomically across processors.
+	AtomicMode Mode = 0xc0
 )
 
+const atomicMask OpCode = 0x0001_ff00
+
+type AtomicOp uint32
+
+const (
+	InvalidAtomic AtomicOp = 0xffff_ffff
+	// AddAtomic - add src to memory address dst atomically
+	AddAtomic AtomicOp = AtomicOp(Add) << 8
+	// AndAtomic - bitwise AND src with memory address at dst atomically
+	AndAtomic AtomicOp = AtomicOp(And) << 8
+	// OrAtomic - bitwise OR src with memory address at dst atomically
+	OrAtomic AtomicOp = AtomicOp(Or) << 8
+	// XorAtomic - bitwise XOR src with memory address at dst atomically
+	XorAtomic AtomicOp = AtomicOp(Xor) << 8
+
+	// LdAcqAtomic - atomically load with acquire semantics
+	LdAcqAtomic AtomicOp = 0x0001_0000
```

Comment:
Maybe we just go with `LoadAcquire` and `StoreRelease` since we don't need to distinguish from Add and Xor, etc?

Reply:
I guess the main issue is the type. If we do not define them, then the user must. And not all ALU operations are supported. Seem more user friendly to have the `AtomicAdd`, `AtomicAnd`, `AtomicOr` and `AtomicXor`.

I agree we don't need the `Fetch` variants, they were mostly there for the string generation. But we can solve that with some logic.